Support WebAssembly SIMD instructions in BBQ baseline JIT

ddegazio opened the pull request:
#### b61cc9d405cba435a3f44269fe50fa07af4e0fbe
<pre>
Support WebAssembly SIMD instructions in BBQ baseline JIT
<a href="https://bugs.webkit.org/show_bug.cgi?id=252433">https://bugs.webkit.org/show_bug.cgi?id=252433</a>
rdar://105465507

Reviewed by Justin Michaud.

Adds support for WebAssembly SIMD instructions to the new single-pass BBQ baseline JIT.
Mostly mimics the Air implementation, but does the Air lowering macros inline since we
directly call out to the macro assembler.

The new baseline JIT is still disabled by default, since it fails quite a few of our
stress tests. It&apos;s not yet ready for typical usage. But with this patch, we at least
expect that the SIMD (and core) spec tests will pass on ARM64 and x86_64.

* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::Value::isFloat const):
(JSC::Wasm::BBQJIT::ControlData::endBlock):
(JSC::Wasm::BBQJIT::logInstruction):
(JSC::Wasm::BBQJIT::loadOpForTypeKind):
(JSC::Wasm::BBQJIT::storeOpForTypeKind):
(JSC::Wasm::BBQJIT::materializePointer):
(JSC::Wasm::BBQJIT::emitLoadOp):
(JSC::Wasm::BBQJIT::emitStoreOp):
(JSC::Wasm::BBQJIT::addSelect):
(JSC::Wasm::BBQJIT::binary):
(JSC::Wasm::BBQJIT::unary):
(JSC::Wasm::BBQJIT::addTopLevel):
(JSC::Wasm::BBQJIT::returnValuesFromCall):
(JSC::Wasm::BBQJIT::addCrash):
(JSC::Wasm::BBQJIT::addSIMDLoad):
(JSC::Wasm::BBQJIT::addSIMDStore):
(JSC::Wasm::BBQJIT::addSIMDSplat):
(JSC::Wasm::BBQJIT::addSIMDShuffle):
(JSC::Wasm::BBQJIT::addSIMDShift):
(JSC::Wasm::BBQJIT::addSIMDExtmul):
(JSC::Wasm::BBQJIT::addSIMDLoadSplat):
(JSC::Wasm::BBQJIT::addSIMDLoadLane):
(JSC::Wasm::BBQJIT::addSIMDStoreLane):
(JSC::Wasm::BBQJIT::addSIMDLoadExtend):
(JSC::Wasm::BBQJIT::addSIMDLoadPad):
(JSC::Wasm::BBQJIT::materializeVectorConstant):
(JSC::Wasm::BBQJIT::addConstant):
(JSC::Wasm::BBQJIT::addExtractLane):
(JSC::Wasm::BBQJIT::addReplaceLane):
(JSC::Wasm::BBQJIT::addSIMDI_V):
(JSC::Wasm::BBQJIT::addSIMDV_V):
(JSC::Wasm::BBQJIT::addSIMDBitwiseSelect):
(JSC::Wasm::BBQJIT::addSIMDRelOp):
(JSC::Wasm::BBQJIT::emitVectorMul):
(JSC::Wasm::BBQJIT::fixupOutOfBoundsIndicesForSwizzle):
(JSC::Wasm::BBQJIT::addSIMDV_VV):
(JSC::Wasm::BBQJIT::dump):
(JSC::Wasm::BBQJIT::emitStore):
(JSC::Wasm::BBQJIT::emitMoveMemory):
(JSC::Wasm::BBQJIT::emitMoveRegister):
(JSC::Wasm::BBQJIT::emitLoad):
(JSC::Wasm::BBQJIT::allocateWithHint):
(JSC::Wasm::BBQJIT::allocateRegister):
(JSC::Wasm::BBQJIT::bind):
(JSC::Wasm::BBQJIT::unbind):
(JSC::Wasm::BBQJIT::evictFPR):
(JSC::Wasm::BBQJIT::clobber):
(JSC::Wasm::BBQJIT::ScratchScope::bindFPRToScratch):
(JSC::Wasm::BBQJIT::ScratchScope::unbindFPRFromScratch):
(JSC::Wasm::BBQJIT::ScratchScope::initializedPreservedSet):

Canonical link: <a href="https://commits.webkit.org/260481@main">https://commits.webkit.org/260481@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/23fde1d8a3ef4006567c42d232ae4d45e2a8e551

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108440 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17536 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41290 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117545 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/116904 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/18987 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8812 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100665 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114208 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/14227 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97445 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/42195 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/96205 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/29090 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/83894 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/97596 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10353 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30434 "Passed tests") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/38/builds/96991 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/8472 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11105 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/7348 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/38/builds/96991 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16500 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50030 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/106017 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7247 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12691 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26236 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -4834,12 +4841,18 @@ class BBQJIT {
+            case TypeKind::V128: {
+                Address address = m_locals[i].asAddress();
+                m_jit.store64(Imm64(0), address);
+                address.offset += 8;
```
IS the mutation here intentional? Can we just store128?

ddegazio replied:
Oh, I guess we totally can store a vector here, I've changed it to a `vectorXor` plus a `storeVector`.

justinmichaud commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -5202,7 +5215,7 @@ class BBQJIT {
                         m_gprSet.add(returnLocation.asGPR(), Width::Width64);
                     else
-                        m_fprSet.add(returnLocation.asFPR(), Width::Width64);
+                        m_fprSet.add(returnLocation.asFPR(), Width::Width128);
```
IgnoreVectors (so 32 bit arm works too) instead of an explicit Width64/128

ddegazio replied:
The new baseline is disabled on 32-bit ARM currently, so this is unnecessary for now. Let's revisit this though, I want to be clear on exactly when we need to be explicit and when `IgnoreVectors` is actually the right move.

justinmichaud commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -5874,7 +6930,7 @@ class BBQJIT {
     Location allocateRegister(TypeKind type)
     {
-        if (type == TypeKind::F32 || type == TypeKind::F64)
+        if (type == TypeKind::F32 || type == TypeKind::F64 || type == TypeKind::V128)
```
isFloat()? BankForReg()?

ddegazio replied:
Resolved in latest push.

justinmichaud commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -5605,6 +6640,21 @@ class BBQJIT {
             m_jit.storePtr(m_dataScratchGPR, dst.asAddress());
             break;
+        case TypeKind::V128: {
+            // We do two scalar load/store pairs instead of one vector load/store in order to avoid
```
Why can't we just clobber FPR scratch? I can't find where we need fprScratch to be live across a move here, where does that happen? Also, why doesn't this switch support doubles?

ddegazio replied:
Seems like it's fine for us to clobber FPR at the moment - there's a few locations where we can get subtle bugs if we remove `m_dataScratchGPR` in general though. I've switched it to use `m_scratchFPR`, and it doesn't seem to be a regression.

justinmichaud reviewed (approved):
r=me

ddegazio committed:
Support WebAssembly SIMD instructions in BBQ baseline JIT
https://bugs.webkit.org/show_bug.cgi?id=252433
rdar://105465507

Reviewed by Justin Michaud.

Adds support for WebAssembly SIMD instructions to the new single-pass BBQ baseline JIT.
Mostly mimics the Air implementation, but does the Air lowering macros inline since we
directly call out to the macro assembler.

The new baseline JIT is still disabled by default, since it fails quite a few of our
stress tests. It's not yet ready for typical usage. But with this patch, we at least
expect that the SIMD (and core) spec tests will pass on ARM64 and x86_64.

* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::Value::isFloat const):
(JSC::Wasm::BBQJIT::ControlData::endBlock):
(JSC::Wasm::BBQJIT::logInstruction):
(JSC::Wasm::BBQJIT::loadOpForTypeKind):
(JSC::Wasm::BBQJIT::storeOpForTypeKind):
(JSC::Wasm::BBQJIT::materializePointer):
(JSC::Wasm::BBQJIT::emitLoadOp):
(JSC::Wasm::BBQJIT::emitStoreOp):
(JSC::Wasm::BBQJIT::addSelect):
(JSC::Wasm::BBQJIT::binary):
(JSC::Wasm::BBQJIT::unary):
(JSC::Wasm::BBQJIT::addTopLevel):
(JSC::Wasm::BBQJIT::returnValuesFromCall):
(JSC::Wasm::BBQJIT::addCrash):
(JSC::Wasm::BBQJIT::addSIMDLoad):
(JSC::Wasm::BBQJIT::addSIMDStore):
(JSC::Wasm::BBQJIT::addSIMDSplat):
(JSC::Wasm::BBQJIT::addSIMDShuffle):
(JSC::Wasm::BBQJIT::addSIMDShift):
(JSC::Wasm::BBQJIT::addSIMDExtmul):
(JSC::Wasm::BBQJIT::addSIMDLoadSplat):
(JSC::Wasm::BBQJIT::addSIMDLoadLane):
(JSC::Wasm::BBQJIT::addSIMDStoreLane):
(JSC::Wasm::BBQJIT::addSIMDLoadExtend):
(JSC::Wasm::BBQJIT::addSIMDLoadPad):
(JSC::Wasm::BBQJIT::materializeVectorConstant):
(JSC::Wasm::BBQJIT::addConstant):
(JSC::Wasm::BBQJIT::addExtractLane):
(JSC::Wasm::BBQJIT::addReplaceLane):
(JSC::Wasm::BBQJIT::addSIMDI_V):
(JSC::Wasm::BBQJIT::addSIMDV_V):
(JSC::Wasm::BBQJIT::addSIMDBitwiseSelect):
(JSC::Wasm::BBQJIT::addSIMDRelOp):
(JSC::Wasm::BBQJIT::emitVectorMul):
(JSC::Wasm::BBQJIT::fixupOutOfBoundsIndicesForSwizzle):
(JSC::Wasm::BBQJIT::addSIMDV_VV):
(JSC::Wasm::BBQJIT::dump):
(JSC::Wasm::BBQJIT::emitStore):
(JSC::Wasm::BBQJIT::emitMoveMemory):
(JSC::Wasm::BBQJIT::emitMoveRegister):
(JSC::Wasm::BBQJIT::emitLoad):
(JSC::Wasm::BBQJIT::allocateWithHint):
(JSC::Wasm::BBQJIT::allocateRegister):
(JSC::Wasm::BBQJIT::bind):
(JSC::Wasm::BBQJIT::unbind):
(JSC::Wasm::BBQJIT::evictFPR):
(JSC::Wasm::BBQJIT::clobber):
(JSC::Wasm::BBQJIT::ScratchScope::bindFPRToScratch):
(JSC::Wasm::BBQJIT::ScratchScope::unbindFPRFromScratch):
(JSC::Wasm::BBQJIT::ScratchScope::initializedPreservedSet):

Canonical link: https://commits.webkit.org/260481@main
